x-pack/plugins/file_upload/public/components/json_upload_and_parse.js

```diff
@@ -0,0 +1,238 @@
+  useEffect(() => {
+    if (!selectedIndexType && indexTypes.length) {
+      setSelectedIndexType(indexTypes[0]);
+    }
+  }, [selectedIndexType, indexTypes]);
+
+  // Set index ready
+  useEffect(() => {
+    const boolIndexReady = !!parsedFile && !!selectedIndexType &&
+      !!indexName && !hasIndexErrors && !indexRequestInFlight;
+    setIndexReady(boolIndexReady);
+    onIndexReadyStatusChange && onIndexReadyStatusChange(boolIndexReady);
+  }, [
+    parsedFile, selectedIndexType, indexName, hasIndexErrors,
+    indexRequestInFlight, onIndexReadyStatusChange
+  ]);
+
+  // Index data
+  useEffect(() => {
+    const filesAreEqual = _.isEqual(indexedFile, parsedFile);
+    if (!boolIndexData || filesAreEqual || !indexReady || indexRequestInFlight) {
+      return;
+    }
+    setIndexRequestInFlight(true);
+
+    indexData(
+      parsedFile, transformDetails, indexName, selectedIndexType, appName
+    ).then(resp => {
+      setIndexedFile(parsedFile);
+      setIndexDataResp(resp);
+      onIndexAddSuccess && onIndexAddSuccess(resp);
+    }).catch(err => {
+      setIndexedFile(null);
+      setIndexDataResp(err);
+      onIndexAddError && onIndexAddError(err);
+    });
+
+  }, [
+    selectedIndexType, boolIndexData, parsedFile, indexedFile, transformDetails,
+    indexName, onIndexAddSuccess, onIndexAddError, appName, indexReady,
+    indexRequestInFlight
+  ]);
+
+  // Index data request complete
+  useEffect(() => {
+    setIndexRequestInFlight(false);
+    if (!boolCreateIndexPattern) {
+      resetFileAndIndexSettings();
+    }
+  }, [indexedFile, boolCreateIndexPattern]);
+
+  // Create Index Pattern
+  useEffect(() => {
+    const indexPatternReady = boolCreateIndexPattern && !!indexedFile &&
+      indexName && !indexPatternRequestInFlight;
+    if (!indexPatternReady) {
+      return;
+    }
+
+    setIndexPatternRequestInFlight(true);
+    createIndexPattern(indexName)
+      .then(resp => {
+        setIndexPatternResp(resp);
+        onIndexPatternCreateSuccess && onIndexPatternCreateSuccess(resp);
+        setIndexPatternRequestInFlight(false);
+      }).catch(err => {
+        setIndexPatternResp(err);
+        onIndexPatternCreateError && onIndexPatternCreateError(err);
+        setIndexPatternRequestInFlight(false);
+      });
+    resetFileAndIndexSettings();
+  }, [
+    indexName, onIndexPatternCreateError, onIndexPatternCreateSuccess,
+    boolCreateIndexPattern, indexPatternRequestInFlight, indexedFile
+  ]);
+
+  // This is mostly for geo. Some data have multiple valid index types that can
+  // be chosen from, such as 'geo_point' vs. 'geo_shape' for point data
+  useEffect(() => {
+    if (parsedFile) {
+      // User-provided index types
+      if (typeof transformDetails === 'object') {
+        setIndexTypes(transformDetails.indexTypes);
+      } else {
+        // Included index types
+        switch(transformDetails) {
+          case 'geo':
+            const featureTypes = _.uniq(
```

Comment:
keep an eye on https://github.com/elastic/kibana/pull/36280, as this introduces the same concept. Once it does, you're likely be able to reuse the logic/constant declarations from that PR

Reply:
Leaving it as is for now since that logic lives in the maps app. This could be rethought in the future though